### PR TITLE
Container MVP usage of config structs

### PR DESCRIPTION
**Pull request by lucksus**
# Container executable
..that creates Holochain instances from a [config file](https://github.com/holochain/holochain-rust/blob/6dac215b8e2ed2ad7c76e253ce37f82472379f4c/container/example-config/basic.toml).
*No interface yet, so not practically usable yet*

## Changes
* `Container` struct added to `container_api` that manages `Holochain` objects and that can populate itself from a Configuration object
* MVP container [executable](https://github.com/holochain/holochain-rust/blob/6dac215b8e2ed2ad7c76e253ce37f82472379f4c/container/src/main.rs) in `container` crate with an [example config](https://github.com/holochain/holochain-rust/blob/6dac215b8e2ed2ad7c76e253ce37f82472379f4c/container/example-config/basic.toml)
* `Configuration::consistency_check()` added, that checks if referenced ids exist
* `Configuration::*_by_id(String)` to easily access elements by id

## Run container exectuable

Container can be run with: `cargo run -- -c example-config/basic.toml `

Full output:
```
lucksus at lucksus15 in ~/code/holochain-rust on container-load-config*
$ cd container
lucksus at lucksus15 in ~/code/holochain-rust/container on container-load-config*
$ cargo run -- -c example-config/basic.toml 
Using config path: example-config/basic.toml
Starting actor system: System[riker]
2018-11-09 23:35:12+00:00 DEBUG [riker::system::system] Actor system [05233d0f-8b71-4b8f-9913-4b00734b2965] [riker] started
Successfully loaded 1 instance configurations
Starting all of them...
Starting instance "app spec instance"...
ok
Done.
^C
lucksus at lucksus15 in ~/code/holochain-rust/container on container-load-config*
$ 
```
You can see that it does run the instance but there is no way yet to interact with it since there is no interface implemented yet. That will be my next PR.

## Upcoming in future PRs
...to get to a full Container MVP
* (memory storage)
* **websocket interface**
* working logger
* full Container API with functions like `install_dna(id: String, path: String)` that change and write config automatically

### CODE REVIEW

File: container_api/src/config.rs [hunk 1]
```diff
@@ -9,17 +10,85 @@ use std::{convert::TryFrom, fs::File, io::prelude::*};
...
+                for ref instance in interface.instances.iter() {
+                    self.instance_by_id(&instance.id).is_some().ok_or_else(|| {
+                        format!(
+                            "Instance configuration \"{}\" not found, mentioned in interface",
+                            instance.id
+                        )
+                    })?;
+                }
+            }
+        }
```
Comment (maackle) [inline container_api/src/config.rs]: Looks like you're building `Result<(), String>` values but aren't using them?

File: container_api/src/config.rs [hunk 2]
```diff
@@ -100,116 +169,193 @@ where
...
+    path = "app_spec_storage"
+
+    "#;
+        let config = load_configuration::<Configuration>(toml).unwrap();
+
+        assert_eq!(config.check_consistency(), Err("DNA configuration \"WRONG DNA ID\" not found, mentioned in instance \"app spec instance\"".to_string()));
+    }
+
+    #[test]
+    fn test_incosistent_config_interface() {
```
Comment (maackle) [inline container_api/src/config.rs]: incoNsistent

File: container_api/src/container.rs [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+                )
+            })
+            .for_each(|(id, maybe_holochain)| {
+                if maybe_holochain.is_ok() {
+                    self.instances.insert(id.clone(), maybe_holochain.unwrap());
+                } else {
+                    println!(
+                        "Error while trying to create instance \"{}\": {}",
+                        id,
+                        maybe_holochain.err().unwrap()
```
Comment (maackle) [inline container_api/src/container.rs]: `match` seems cleaner rather than `if` + `unwrap`

File: container_api/src/container.rs [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+            let dna = dna_loader((&dna_config.file)).map_err(|_| {
+                HolochainError::ConfigError(format!(
+                    "Could not load DNA file \"{}\"",
+                    dna_config.file
+                ))
+            })?;
+
+            (instance_config.storage.storage_type == "file"
+                && instance_config.storage.path.is_some())
+                .ok_or(String::from("Only file storage supported currently"))?;
```
Comment (maackle) [inline container_api/src/container.rs]: `try!` might actually make this more readable

File: container/Cargo.toml [hunk 5]
```diff
@@ -4,5 +4,6 @@ version = "0.0.1"
 authors = ["Holochain Core Dev Team <devcore@holochain.org>"]
 
 [dependencies]
-holochain_dna = { path = "../dna" }
-holochain_core = { path = "../core" }
\ No newline at end of file
+holochain_core_types = { path = "../core_types" }
+holochain_container_api = { path = "../container_api" }
+clap = "2"
```
Comment (sphinxc0re) [inline container/Cargo.toml]: Missing newline

File: container/example-config/basic.toml [hunk 6]
```diff
@@ -1,27 +1,27 @@
-[[agent]]
+[[agents]]
```
Comment (sphinxc0re) [inline container/example-config/basic.toml]: Why did you change this to plural?

File: container/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+use clap::{App, Arg};
+use holochain_container_api::{
+    config::{load_configuration, Configuration},
+    container::Container,
+};
+use holochain_core_types::error::HolochainError;
+use std::{convert::TryFrom, fs::File, io::prelude::*};
+
+fn main() {
+    let matches = App::new("hcc")
```
Comment (sphinxc0re) [inline container/src/main.rs]: Please use `StructOpt` for the Cli. It avoids duplicating a lot of code and writing a lot of boilerplate

File: container_api/src/container.rs [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+impl Container {
+    pub fn new() -> Self {
+        Container {
+            instances: HashMap::new(),
+        }
+    }
+
+    pub fn start_all(&mut self) {
+        let _ = self.instances.iter_mut().for_each(|(id, hc)| {
+            println!("Starting instance \"{}\"...", id);
```
Comment (zippy) [inline container_api/src/container.rs]: I think we shouldn't really be assuming that the container wants to print stuff.  If anything it should be a log or a console or something like that.  This is fine for now, but we should log a followup (or put it in the tree)

File: container_api/src/container.rs (hunk 4 above)
Comment (lucksus) [reply container_api/src/container.rs]: Linter thinks, ? is cleaner: screenshot 2018-11-12 at 00 03 11

File: container/src/main.rs (hunk 7 above)
Comment (lucksus) [reply container/src/main.rs]: thanks 🙏

File: container/example-config/basic.toml (hunk 6 above)
Comment (lucksus) [reply container/example-config/basic.toml]: Because @maackle convinced me: https://github.com/holochain/holochain-rust/pull/598/files/f1a93b42af7cd2594aa2bfe8736c9ca55c77d86a#r232011685

File: container_api/src/container.rs (hunk 8 above)
Comment (lucksus) [reply container_api/src/container.rs]: Yeah, container_api should maybe not and I will remove this in future PRs. Felt right to have the container executable print to std-out which can be redirected by the shell context that runs the process.

File: container_api/src/config.rs (hunk 1 above)
Comment (lucksus) [reply container_api/src/config.rs]: Am I? Not sure what you mean. `check_consistency()` and the error message is used here: https://github.com/holochain/holochain-rust/pull/605/files#diff-927e383d6b6393e9f319091b2e4fd147R54

Comment (maackle) [reply container_api/src/config.rs]: Ahhh, sorry I missed the `?` on L47.

**Comment by lucksus** [comment]
@Connoropolous, @sphinxc0re, implemented your change requests.
Please re-review.

File: container_api/src/container.rs [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+    io::prelude::*,
+    sync::{Arc, Mutex, RwLock},
+};
+
+use boolinator::*;
+
+/// Main representation of the container.
+/// Holds a `HashMap` of Holochain instances referenced by ID.
+///
+/// The main juice of this struct is
```
Comment (Connoropolous) [inline container_api/src/container.rs]:
```suggestion
/// A primary point in this struct is
```
